Add generic OIDC provider support

jordojordo opened the pull request:
<!-- This template is for Devs to give QA details before moving the issue To-Test -->
### Summary
Fixes #9820 
<!-- Define findings related to the feature or bug issue. -->

This adds a generic OIDC authentication provider to the auth config view.

### Occurred changes and/or fixed issues
<!-- Include information of the changes, including collateral areas which have been affected by this PR as requirement or for convenience. -->
- Refactored the oicd.vue component for editing the provider to remove Keycloak references as this is now agnostic.
- Added "Advanced Settings" section to include:
  - `jwksUrl`
  - `tokenEndpoint`
  - `userInfoEndpoint`
  - `scope`


### Areas or cases that should be tested
<!-- Areas that should be tested can include Airgap checks, Rancher upgrades, K8s upgrade, etc. -->
<!-- Which browser did you use for local testing? The reviewer should test with a different browser. -->
<!-- Add missing steps or rewrite them if have been missed or to complement existing information. This should define a clear way to reproduce it and not an approximation. -->
A testing environment can be supplied on request.


### Screenshot/Video
<!-- Attach screenshot or video of the changes and eventual comparison if you find it necessary -->

___Note: Excuse the "genericoidc" provider label as this was created from a test build Rancher image___

https://github.com/rancher/dashboard/assets/40806497/4c63c05b-7ca4-4e25-941f-b243a70fc04a



### Checklist
- [ ] The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed
- [ ] The PR has a Milestone <!-- The milestone should automatically be assigned if the linked issue has one, but does not always happen (incorrectly linked, issue has no milestone, etc) -->
- [ ] The PR template has been filled out
- [ ] The PR has been self reviewed <!-- There are no TODOs, no incorrect files in the PR, all the required files are there, no commented out code, etc-->
- [ ] The PR has a reviewer assigned
- [ ] The PR has automated tests or clear instructions for manual tests and the linked issue has appropriate QA labels, or tests are not needed
- [ ] The PR has reviewed with UX and tested in light and dark mode, or there are no UX changes

richard-cox commented on shell/edit/auth/oidc.vue:
```diff
@@ -58,43 +67,64 @@ export default {
+      const url = this.oidcUrls.url.replaceAll(' ', '');
 
-      this.model.issuer = `${ url }/auth/realms/${ this.keycloakUrls.realm || '' }`;
+      this.model.issuer = `${ url }/auth/realms/${ this.oidcUrls.realm || '' }`;
```
In Chad's test instance the OIDC urls don't contain `auth`
- Issuer - `.../auth/realms/testrealm` --> `.../realms/testrealm`
- Auth Endpoint - `..../auth/realms/testrealm/protocol/openid-connect/auth` --> `..../realms/testrealm/protocol/openid-connect/auth`

@crobby what would be sensible defaults to use for generic oidc urls, is your instance typical and the `auth` in the path is more of a keycloak thing?

crobby replied:
I think the paths in my sample keycloak instance are a bit of "a keycloak thing"
I'm not sure what the most sensible default would be.  Let me poke around a bit.

crobby replied:
Maybe it makes the most sense to have no default.  The generic oidc implementation will perform "discovery" based on the Issuer if the other endpoints are not provided, so defaulting it to empty may make the most sense.  From what I've seen, having discovery turned on is the most typical configuration.

richard-cox replied:
@crobby ATM the UI requests the root url and realm, and then constructs mandatory `issuer` and `authEndpoint` urls.  

- We would need at least the issuer url
- Is the authEnpoint mandatory, or does discovery work from just the issuer?

I get the feeling we should just go with the removal of `auth` from the urls. The generate step is designed to get the user quickly to the point of discovery. If they don't work then the user can choose to `Specify` and tweak our suggestions, and then optional avoid discovery by adding more advanced urls

crobby replied:
Auth is not mandatory for the generic provider.  If it's not present, we'll use the issuer to perform discovery for all the other endpoints.
I like the idea of removing auth from the "generate" step.

richard-cox reviewed (changes requested):
- Reach out in slack to Ken / Eden with regards to 
  - a logo to use for open idc (there's a few options)
  - the label we use for oidc (was wondering if instead of `OIDC` we use `Generic OIDC`)
- The PR should explicitly call out possible regressions for the Keycloak OIDC provider. We should also mention enabling, editing and disabling of both oidc providers needs validating  

Update

Chat tweaked his oidc settings and we can now dev against a local rancher with his dev image... and his dev oidc instance. I've tested and can successful configure, log in and 'lot out'. Found the following though
- Once the oidc provider is configured and returning to the auth provider page i cann't change the 'who should be able to login' settings (either mode on left or specific users/groups on right) as the `Save` button is disabled. I think this is due to the new validation

jordojordo commented on shell/edit/auth/oidc.vue:
```diff
@@ -266,13 +303,56 @@ export default {
           </div>
         </div>
+
+        <AdvancedSection :mode="mode">
```
As far as I know they do match, I was able to create a keycloak auth while inputting those advanced settings.

jordojordo commented on shell/utils/auth.js:
```diff
@@ -64,8 +64,6 @@ export function returnTo(opt, vm) {
   const rows = await store.dispatch(`management/findAll`, { type: MANAGEMENT.AUTH_CONFIG });
   const nonLocal = rows.filter((x) => x.name !== 'local');
-  // Generic OIDC is returned via API but not supported (and will be removed or fixed in future)
-  const supportedNonLocal = nonLocal.filter((x) => x.id !== 'oidc');
```
I replaced the filter to remove the `genericoidc` id.

jordojordo commented:
I spoke with Ken and he provided an icon to use, I updated the auth card to use this icon. Also changed the title wording to "Generic OIDC"

![icons](https://github.com/rancher/dashboard/assets/40806497/cbc279f1-f896-4cab-903f-ef64194a1310)

> Once the oidc provider is configured ... and then returning to the auth provider page i can't change the 'who should be able to login' settings (either mode on left or specific users/groups on right) as the Save button is disabled. I think this is due to the new validation

The validation was not checking for the `editConfig` state, I fixed the validation to add this condition.

jordojordo commented on shell/edit/auth/oidc.vue:
```diff
@@ -266,13 +303,56 @@ export default {
+
+          <div class="row mb-20">
+            <div class="col span-6">
+              <ArrayList
```
I changed the inputs to initially apply the `model.scope` now, the inputs will be disabled unless the "Custom Endpoints" checkbox is true.
I don't think it's correct to say they are the minimum required scopes as this is solely dependent on the realm's config.

richard-cox replied:
Interesting, i always thought as the UI requests those three (openid, email, profile) when logging in the request would be rejected if one was not configured however that's not the case. It looks like the scopes are added to the verify url that we expect back.

@crobby Can you confirm how the authconfig scopes are used?

crobby replied:
> Interesting, i always thought as the UI requests those three (openid, email, profile) when logging in the request would be rejected if one was not configured however that's not the case. It looks like the scopes are added to the verify url that we expect back.
> 
> @crobby Can you confirm how the authconfig scopes are used?

The scopes indicate which bits of data we are requesting about the user.  It is possible that Rancher logins will still "work" to some degree with some scopes omitted, but for Rancher to get all of the data we'd like to get, we default to "openid profile email".  I believe that only openid is required as part of an OIDC authentication.

richard-cox replied:
Ok. So...

- Auth Config
  - we should ALWAYS require `openid`
  - we should default to have `profile` and `email` entries
- UI requested scopes (BASE_SCOPES) should contain at least openid (when we don't as per current code the OIDC test when configuring provider fails)

richard-cox commented on shell/assets/translations/en-us.yaml:
```diff
@@ -6483,7 +6491,7 @@ model:
       freeipa: FreeIPA
       googleoauth: Google
-      oidc: OIDC
+      oidc: Generic OIDC
```
Problem compounds. `Generic OIDC` looks good when selecting it as a provider, however looks bad when logging in `Log in with Generic OIDC`. Can the log in button still remain as `Log in with OIDC`? Should be possible via the log in component.

Apologies for the churn.

jordojordo replied:
No worries, I updated the login button to use `OIDC` as the provider display name:

![oidclogin](https://github.com/rancher/dashboard/assets/40806497/4006aaa0-289a-40b6-ba0d-c86782bb8ad6)

richard-cox reviewed (changes requested):
just one small practical thing to change, but a few remaining items to check with Chad

jordojordo commented:
I have moved the Scope inputs above the Endpoints section as these are separate from the endpoint configuration. I could not find a way to have one input within the array list be required/disabled, so I opted for a tooltip explaining the requirement of `openid` in the scope. `openid`, `profile`, and `email` entries are now added by default to the scope.

The `ACR` now has an input within the Endpoints advanced configuration, this adds the `acrValue` property to the model. When testing this was causing issues with the image provided by @crobby, it could be an issue with using the `authconfig/oidc` provider rather than the `authconfig/genericoidc` - but I was unable to confirm this.

richard-cox commented on shell/assets/translations/en-us.yaml:
```diff
@@ -573,11 +573,21 @@ authConfig:
+    scope:
+      label: Scope
+      placeholder: openid
+      protip: The scope parameter requires <code>openid</code> to be present. If you are using a custom scope, ensure that openid is included.
```
```suggestion
      protip: The <code>openid</code> scope is required. If you are using custom scopes, ensure that it is included.
```

there's one scope property, but each entry is commonly known as a scope.

richard-cox commented on shell/assets/translations/en-us.yaml:
```diff
@@ -573,11 +573,21 @@ authConfig:
+    userInfoEndpoint: User Info Endpoint
+    acrValue: Authorization Context Reference
+    scope:
+      label: Scope
```
```suggestion
      label: Scopes
```

Following on from the `protip` comment this should probably be plural

richard-cox commented on shell/edit/auth/oidc.vue:
```diff
@@ -59,42 +68,67 @@ export default {
       };
     },
 
+    validationPassed() {
```
This block could be tidied up, and we shouldn't lose the validation for keycloak

richard-cox commented on shell/edit/auth/oidc.vue:
```diff
@@ -59,42 +68,67 @@ export default {
+      const url = this.oidcUrls.url.replaceAll(' ', '');
 
-      this.model.issuer = `${ url }/auth/realms/${ this.keycloakUrls.realm || '' }`;
+      this.model.issuer = `${ url }/realms/${ this.oidcUrls.realm || '' }`;
```
we shouldn't lose the `auth` for keycloak

richard-cox commented on shell/edit/auth/oidc.vue:
```diff
@@ -154,7 +188,8 @@ export default {
               :label="t(`authConfig.oidc.clientId`)"
               :mode="mode"
-              required
+              requiredoidcScope
```
think there's some rogue c&p here `requiredoidcScope`

richard-cox commented on shell/edit/auth/oidc.vue:
```diff
@@ -59,42 +68,67 @@ export default {
 
-      this.model.issuer = `${ url }/auth/realms/${ this.keycloakUrls.realm || '' }`;
+      this.model.issuer = `${ url }/realms/${ this.oidcUrls.realm || '' }`;
       this.model.authEndpoint = `${ this.model.issuer || '' }/protocol/openid-connect/auth`;
```
following on from https://github.com/rancher/dashboard/pull/11112#discussion_r1633350312, for generic oidc this now isn't mandatory and we shouldn't pre-populate it (which means the backend will do discovery)

richard-cox replied:
Update: I just gave this a try and get the following error for /v3/keyCloakOIDCConfigs/keycloakoidc?action=configureTest

```
{"baseType":"error","code":"NotNullable","fieldName":"authEndpoint","message":"","status":422,"type":"error"}
```

@crobby was there an updated image that supports this? does discovery work if both issuer and authEndpoint are supplied (we're sending an empty string for the later, but not sure that would cause the error)?

crobby replied:
My mistake.  I will push a new image that allows for an empty auth endpoint.

richard-cox commented:
@jordojordo let me know when this one's ready to re-review

jordojordo commented:
@richard-cox This is ready for review now, there were a few hiccups in the test image but those have been fixed.

richard-cox commented on shell/edit/auth/oidc.vue:
```diff
@@ -59,42 +68,74 @@ export default {
+      }
+
+      const { clientId, clientSecret } = this.model;
+      const isValidScope = this.model.id === 'keycloakoidc' || this.oidcScope?.includes('openid');
```
This could be optimised / clearer by exiting early if this condition is false (as it's in both return statements below). Doesn't need changing now though

richard-cox commented on shell/edit/auth/oidc.vue:
```diff
@@ -257,22 +317,64 @@ export default {
               v-model="model.authEndpoint"
               :label="t(`authConfig.oidc.authEndpoint`)"
               :mode="mode"
-              required
```
auth endpoint is still required for keycloak oidc

richard-cox reviewed (changes requested):
The PR should explicitly call out possible regressions for the Keycloak OIDC provider. We should also mention that enabling, editing and disabling of both oidc providers needs validating. **This information is missing from PR description in the `Areas or cases that should be tested` section.**
